packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -521,60 +621,145 @@ add (const Scalar& alpha,
       (!transposeA &&  transposeB && !A.getRangeMap ()->isSameAs (*B.getDomainMap())) ||
       ( transposeA && !transposeB && !A.getDomainMap()->isSameAs (*B.getRangeMap ()));
     TEUCHOS_TEST_FOR_EXCEPTION(rangeMapsSame, std::invalid_argument,
-      prefix << "The range Maps of Op(A) and Op(B) are not the same.");
+      prefix_mmm << "The range Maps of Op(A) and Op(B) are not the same.");
   }
 #endif // HAVE_TPETRA_DEBUG
-
+  auto comm = A.getComm();
   // Form the explicit transpose of A if necessary.
-  RCP<const crs_matrix_type> Aprime;
+  RCP<const crs_matrix_type> Aprime = rcpFromRef(A);
   if (transposeA) {
-    transposer_type transposer (rcpFromRef (A));
+    transposer_type transposer(Aprime);
     Aprime = transposer.createTranspose ();
-
-  } else {
-    Aprime = rcpFromRef (A);
   }
-
 #ifdef HAVE_TPETRA_DEBUG
   TEUCHOS_TEST_FOR_EXCEPTION(Aprime.is_null (), std::logic_error,
-    prefix << "Failed to compute Op(A). Please report this bug to the Tpetra developers.");
+    prefix_mmm << "Failed to compute Op(A). Please report this bug to the Tpetra developers.");
 #endif // HAVE_TPETRA_DEBUG
-
   // Form the explicit transpose of B if necessary.
-  RCP<const crs_matrix_type> Bprime;
+  RCP<const crs_matrix_type> Bprime = rcpFromRef(B);
   if (transposeB) {
-    transposer_type transposer (rcpFromRef (B));
+    transposer_type transposer(Bprime);
     Bprime = transposer.createTranspose ();
-
-  } else {
-    Bprime = rcpFromRef (B);
   }
-
 #ifdef HAVE_TPETRA_DEBUG
   TEUCHOS_TEST_FOR_EXCEPTION(Bprime.is_null (), std::logic_error,
-    prefix << "Failed to compute Op(B). Please report this bug to the Tpetra developers.");
-
+    prefix_mmm << "Failed to compute Op(B). Please report this bug to the Tpetra developers.");
   TEUCHOS_TEST_FOR_EXCEPTION(
     !Aprime->isFillComplete () || !Bprime->isFillComplete (), std::invalid_argument,
-    prefix << "Aprime and Bprime must both be fill complete.  "
+    prefix_mmm << "Aprime and Bprime must both be fill complete.  "
     "Please report this bug to the Tpetra developers.");
 #endif // HAVE_TPETRA_DEBUG
-
-  RCP<row_matrix_type> C =
-    Bprime->add (alpha, *rcp_implicit_cast<const row_matrix_type> (Aprime),
-                 beta, domainMap, rangeMap, params);
-
-  return rcp_dynamic_cast<crs_matrix_type> (C);
+  RCP<const map_type> CDomainMap = domainMap;
+  RCP<const map_type> CRangeMap = rangeMap;
+  if(CDomainMap.is_null())
+  {
+    CDomainMap = Bprime->getDomainMap();
+  }
+  if(CRangeMap.is_null())
+  {
+    CRangeMap = Bprime->getRangeMap();
+  }
+  assert(!(CDomainMap.is_null()));
+  assert(!(CRangeMap.is_null()));
+  typedef typename AddKern::ValuesArray ValuesArray;
+  typedef typename AddKern::RowPtrsArray RowPtrsArray;
+  typedef typename AddKern::ColIndsArray ColIndsArray;
+  bool AGraphSorted = Aprime->getCrsGraph()->isSorted();
+  bool BGraphSorted = Bprime->getCrsGraph()->isSorted();
+  RCP<ValuesArray> vals;
```

Comment:
See above.  Kokkos::View is a lot like Teuchos::ArrayRCP.  You wouldn't generally have an `RCP<ArrayRCP>`, would you?